hip/components/memory.hip.hpp

```diff
@@ -20,71 +21,203 @@ namespace kernels {
 namespace hip {
 
 
-#include "common/cuda_hip/components/memory.hpp.inc"
+#if GINKGO_HIP_PLATFORM_NVCC
 
 
-template <typename ValueType>
+#include "common/cuda_hip/components/memory.nvidia.hpp.inc"
+
+
+#else
+
+
+template <typename T>
+struct gcc_atomic_intrinsic_type_map {};
+
+
+template <>
+struct gcc_atomic_intrinsic_type_map<int32> {
+    using type = int32;
+};
+
+
+template <>
+struct gcc_atomic_intrinsic_type_map<float> {
+    using type = int32;
+};
+
+
+template <>
+struct gcc_atomic_intrinsic_type_map<int64> {
+    using type = int64;
+};
+
+
+template <>
+struct gcc_atomic_intrinsic_type_map<double> {
+    using type = int64;
+};
+
+
+#if HIP_VERSION >= 50100000
+#define HIP_ATOMIC_LOAD(ptr, memorder, scope) \
+    __hip_atomic_load(ptr, memorder, scope)
+#define HIP_ATOMIC_STORE(ptr, value, memorder, scope) \
+    __hip_atomic_store(ptr, value, memorder, scope)
+#define HIP_SCOPE_GPU __HIP_MEMORY_SCOPE_AGENT
+#define HIP_SCOPE_THREADBLOCK __HIP_MEMORY_SCOPE_WORKGROUP
+#else
+#define HIP_ATOMIC_LOAD(ptr, memorder, scope) __atomic_load_n(ptr, memorder)
+#define HIP_ATOMIC_STORE(ptr, value, memorder, scope) \
+    __atomic_store_n(ptr, value, memorder)
+#define HIP_SCOPE_GPU -1
+#define HIP_SCOPE_THREADBLOCK -1
+#endif
+
+
+template <int memorder, int scope, typename ValueType>
+__device__ __forceinline__ ValueType load_generic(const ValueType* ptr)
+{
+    using atomic_type = typename gcc_atomic_intrinsic_type_map<ValueType>::type;
+    static_assert(sizeof(atomic_type) == sizeof(ValueType), "invalid map");
+    static_assert(alignof(atomic_type) == sizeof(ValueType), "invalid map");
+    auto cast_value = HIP_ATOMIC_LOAD(reinterpret_cast<const atomic_type*>(ptr),
+                                      memorder, scope);
+    ValueType result{};
+    memcpy(&result, &cast_value, sizeof(ValueType));
```

Comment:
`reinterpret_cast` the result to the `atomic_type`, too?

Reply:
`memcpy` uses `void*`, so it doesn't matter

Reply:
good to hear they give the same codegen in the end. Thanks for trying both